src/dmd/expressionsem.d

```diff
@@ -8757,22 +8757,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                         result = e;
                         return;
                     }
+                    // https://issues.dlang.org/show_bug.cgi?id=21586
+                    // Rewrite CondExp or e1 will miss direct construction, e.g.
+                    // e1 = a ? S(1) : ...; -> AST: e1 = a ? (S(0)).this(1) : ...;
+                    // a temporary created and an extra destructor call.
+                    // AST will be rewritten to:
+                    // a ? e1 = 0, e1.this(1) : ...; -> blitting plus construction
+                    if (e2x.op == TOK.question)
+                    {
+                        /* Rewrite as:
+                         *  a ? e1 = b : e1 = c;
```

Comment:
Does this also change the evaluation order? `myDest() = myCond() ? a : b;` -> `myDest` must still be called before `myCond`.

Reply:
But isn't that an `AssignExp`? Here we are inside `ConstructExp` logic.

Reply:
Right; I guess it's unlikely `e1` has side effects, but it probably cannot be ruled out.